File: GreedyStrategy.rb
require_relative 'HangmanGame'
require_relative 'Guess'
require_relative 'GuessingStrategy'
require_relative 'GuessLetter'
require_relative 'GuessWord'
require_relative 'Dictionary'

class GreedyStrategy < GuessingStrategy

  def initialize(dictionary)
    @dictionary = dictionary
    # strategy needs access to processed list of all possible words
  end

  def nextGuess(game)

    if game.currentScore == 0
      # no guesses have been made yet
      # start possible words with words of right length
      @possible_words += @dictionary.word_length[game.getSecretWordLength]
      raise "No possible words" if @possible_words.empty?
      return make_guess(game)
    end

    # only filter out letters that haven't already been filtered
    new_correct_letter = (game.getCorrectlyGuessedLetters - @correctly_guessed_letters).first
    if new_correct_letter
      (0...(game.getSecretWordLength)).each do |i|
        if game.getGuessedSoFar[i] == new_correct_letter
          # only keep words that have the letter in the right position
          position_matching = @dictionary.getWordsWithLetterAtPosition(new_correct_letter, i)
          raise "No possible words" if position_matching.nil?
          @possible_words = @possible_words & position_matching
        end

        if game.getGuessedSoFar[i] == HangmanGame::MYSTERY_LETTER
          # remove words that have letter in non-revealed position
          position_mismatch = @dictionary.getWordsWithLetterAtPosition(new_correct_letter, i)
          @possible_words = @possible_words - position_mismatch
        end
      end

      @correctly_guessed_letters << new_correct_letter
      return make_guess(game)
    end

    # remove words containing wrong letters
    new_wrong_letter = (game.getIncorrectlyGuessedLetters - @incorrectly_guessed_letters).first
    if new_wrong_letter
      @possible_words = @possible_words - @dictionary.getWordsWithLetter(new_wrong_letter)
      @incorrectly_guessed_letters << new_wrong_letter
      return make_guess(game)
    end

    # remove wrong words
    new_wrong_word = (game.getIncorrectlyGuessedWords - @incorrectly_guessed_words).first
    if new_wrong_word
      @possible_words.delete(@dictionary.find(new_wrong_word))
      @incorrectly_guessed_words << new_wrong_word
      return make_guess(game)
    end

    # one of the 4 if statements should be the only 4 possible outcomes
    raise 'No changes detected in game :('
    return make_guess(game, @possible_words)
  end #nextGuess

  def make_guess(game)
    case @possible_words.size
      when 1,2
        return GuessWord.new(@possible_words.first.word)
      else
        frequency = calc_letter_frequency(@possible_words, game.getAllGuessedLetters)
        return GuessLetter.new(pick_most_used(frequency))
    end
  end

  def newGame
    # call at beginning of run()
      @possible_words = Set.new
      @incorrectly_guessed_words = Set.new
      @correctly_guessed_letters = Set.new
      @incorrectly_guessed_letters = Set.new
  end

  def calc_letter_frequency(words, used_letters)
    frequency = {}
    words.each do |word|
      word.unique_letters.each do |ch|
        if frequency[ch].nil?
          frequency[ch] = 1
        else
          frequency[ch] += 1
        end
      end
    end

    used_letters.each do |ch|
      frequency.delete(ch)
    end

    frequency
  end

  def pick_most_used(frequency)
    max = 0
    char = ''
    frequency.each do |key,value|
      if value > max
        char = key
        max = value
      end
    end
    char
  end # pick_char

end

File: GreedyStrategy.java

import java.util.*;

public class GreedyStrategy implements GuessingStrategy {

    private Set<Word> possibleWords;
    private Set<Character> incorrectlyGuessedLetters;
    private Set<Character> correctlyGuessedLetters;
    private Set<String> incorrectlyGuessedWords;

    public final Dictionary dictionary;

    public GreedyStrategy(Dictionary d) {
        dictionary = d;
        // strategy needs dictionary access to processed list of all possible words
    }

    @Override
    public Guess nextGuess(HangmanGame game) {

        if (game.currentScore() == 0) {
            // no guesses have been made yet
            // start possible words with words of right length
            possibleWords.addAll(dictionary.getWordsWithLength(game.getSecretWordLength()));
            if (possibleWords.isEmpty()) {
                throw new RuntimeException("No possible words");
            }
            return makeGuess(game);
        }

        // only filter out letters that haven't already been filtered
        Set<Character> newCorrectlyGuessedLetters = new HashSet<>(game.getCorrectlyGuessedLetters());
        newCorrectlyGuessedLetters.removeAll(correctlyGuessedLetters);
        if (newCorrectlyGuessedLetters.size() > 0) {
            // only keep words that have the letter in the right position
            Character newCorrectLetter = getOnlyChar(newCorrectlyGuessedLetters);
            for (int i=0; i<game.getSecretWordLength(); i++) {
                if (game.getGuessedSoFar().charAt(i) == newCorrectLetter) {
                    Set<Word> positionMatching = dictionary.getWordsWithLetterAtPosition(newCorrectLetter, i);
                    possibleWords.retainAll(positionMatching);
                }
                
                // remove words that have letter in non-revealed position
                if (game.getGuessedSoFar().charAt(i) == HangmanGame.MYSTERY_LETTER) {
                    Set<Word> positionMismatch = dictionary.getWordsWithLetterAtPosition(newCorrectLetter, i);
                    possibleWords.removeAll(positionMismatch);
                }
            }
            
            correctlyGuessedLetters.add(newCorrectLetter);
            return makeGuess(game);
        }
        
        // remove words containing wrong letters
        Set<Character> newWrongLetters = new HashSet<>(game.getIncorrectlyGuessedLetters());
        newWrongLetters.removeAll(incorrectlyGuessedLetters);
        if (newWrongLetters.size() > 0) {
            Character newWrongLetter = getOnlyChar(newWrongLetters);
            possibleWords.removeAll(dictionary.getWordsWithLetter(newWrongLetter));
            incorrectlyGuessedLetters.add(newWrongLetter);
            return makeGuess(game);
        }
        
        // remove wrong words
        Set<String> newWrongWords = new HashSet<>(game.getIncorrectlyGuessedWords());
        newWrongWords.removeAll(incorrectlyGuessedWords);
        if (newWrongWords.size() > 0) {
            String newWrongWord = getOnlyWord(newWrongWords);
            possibleWords.remove(dictionary.find(newWrongWord));
            incorrectlyGuessedWords.add(newWrongWord);
            return makeGuess(game);
        }
        
        // one of the 4 if statements should be the only 4 possible outcomes
        throw new RuntimeException("No changes detected in game");
    }

    public void newGame() {
        // call at beginning of run        
        possibleWords = new HashSet<>();
        incorrectlyGuessedLetters = new HashSet<>();
        correctlyGuessedLetters = new HashSet<>();
        incorrectlyGuessedWords = new HashSet<>();
    }

    private String getOnlyWord(Set<String> set) {
        // quite unsafe/dirty method of getting the remaining word in a set
        if (set.size() == 1) {
            Iterator<String> iter = set.iterator();
            if (iter.hasNext()) {
                return iter.next();
            } else {
                throw new RuntimeException("Something went very wrong here.");
            }
        } else {
            throw new RuntimeException("Set has more than one element.");
        }
    }

    private Character getOnlyChar(Set<Character> set) {
        // quite unsafe/dirty method of getting the remaining char in a set
        if (set.size() == 1) {
            Iterator<Character> iter = set.iterator();
            if (iter.hasNext()) {
                return iter.next();
            } else {
                throw new RuntimeException("Something went very wrong here.");
            }
        } else {
            throw new RuntimeException("Set has more than one element.");
        }
    }
    
    private HashMap<Character, Integer> CalcLetterFrequency(Set<Word> words, Set<Character> usedLetters) {
        HashMap<Character, Integer> frequency = new HashMap<>();

        for (Word word : words) {
            for (Character ch : word.uniqueLetters) {
                if (frequency.get(ch) == null) {
                    frequency.put(ch, 1);
                } else {
                    frequency.put(ch, frequency.get(ch)+1);
                }
            }
        }
        
        for (Character ch : usedLetters) {
            frequency.remove(ch);
        }
       
        return frequency;
    }
    
    private Character pickMostUsed(Map<Character, Integer> frequency) {
        Integer max = 0;
        Character ch = 'E';
        for (Map.Entry<Character, Integer> entry : frequency.entrySet()) {
            Character key = entry.getKey();
            Integer value = entry.getValue();
            if (value > max) {
                ch = key;
                max = value;
            }
        }
        return ch;
    }
    
    

    private Guess makeGuess(HangmanGame game) {
        switch(possibleWords.size()) {
            case 1:
            case 2:
                // pick the only remaining word, or pick one of the two at random
                Iterator<Word> iter = possibleWords.iterator();
                return new GuessWord(iter.next().word);
            default:
                Map<Character, Integer> frequency = CalcLetterFrequency(possibleWords, game.getAllGuessedLetters());
                return new GuessLetter(pickMostUsed(frequency));
        }
    }

}
